src/main/java/technology/semi/weaviate/client/v1/auth/provider/AuthTokenProvider.java

```diff
@@ -0,0 +1,39 @@
+package technology.semi.weaviate.client.v1.auth.provider;
+
+import java.util.TimerTask;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import org.apache.commons.lang3.StringUtils;
+import technology.semi.weaviate.client.Config;
+
+public class AuthTokenProvider extends NimbusAuth implements AccessTokenProvider {
+  private String accessToken;
+  private ScheduledExecutorService executor;
+
+  public AuthTokenProvider(Config config, AuthResponse authResponse, String accessToken, long lifetimeSeconds, String refreshToken) {
+    this.accessToken = accessToken;
+    if (StringUtils.isNotBlank(refreshToken)) {
+      scheduleRefreshTokenTask(config, authResponse, refreshToken, lifetimeSeconds);
+    }
+  }
+
+  @Override
+  public String getAccessToken() {
+    return accessToken;
+  }
+
+  private void scheduleRefreshTokenTask(Config config, AuthResponse authResponse, String refreshToken, long period) {
+    TimerTask refreshTokenTask = new TimerTask() {
+      public void run() {
+        accessToken = refreshToken(config, authResponse, refreshToken);
+      }
+    };
+    executor = Executors.newSingleThreadScheduledExecutor();
```

Comment:
are there any tests that make sure that this works?

There are two cases that would need to be tested, ideally with a server mock:
- User/Pw (cannot be saved => refresh with access/refresh token)
- CC => some auth providers (azure/okta, however keycloak returns one) do not return a refresh token, so the client credentials need to be saved in the client to allow fetching a new token.

Reply:
I have added and `e2e` test for `refreshToken` that uses our `WCS`

Reply:
It's called `NimbusAuthRefreshTokenTest`

Reply:
cool! :)